File: Python/32. Longest Valid Parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        result: int = 0
        array = [-1]

        for i in range(len(s)):
            if s[i] == '(':
                array.append(i)
            else:
                array.pop()
                if not array:
                    array.append(i)
                else:
                    result = max(result, i - array[-1])
        return result

if __name__ == "__main__":
    solution = Solution()
    # For testing
    # long = solution.longestValidParentheses("))))())()()(()")
    # print(long)

File: Javascript/32. Longest Valid Parentheses.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    let result = 0
    let array = [-1]

    for (let i = 0; i < s.length; i++) {
        if (s[i] === "(") {
            array.push(i)
        } else {
            array.pop()
            if (array.length === 0) {
                array.push(i)
            } else {
                result = Math.max(result, i - array[array.length-1])
            }
        }
    }
    return result
};

// longestValidParentheses("))))())()()(()")